src/AutomaticScheduling.cpp

```diff
@@ -0,0 +1,255 @@
+        if (dim == 0) {
+            // Unable to compute the bounds for the function. Return undefined bounds.
+            result.clear();
+            return result;
+        }
+        internal_assert((int)dim == f.dimensions());
+        for (unsigned i = 0; i < dim; ++i) {
+            Interval interval = b.bounds[i];
+            if (!initialized[i]) {
+                initialized[i] = true;
+                result[i] = interval;
+            } else {
+                result[i].min = min(result[i].min, interval.min);
+                result[i].max = max(result[i].max, interval.max);
+            }
+        }
+    }
+    return result;
+}
+
+
+/** Return the footprint (required region) of the given function over
+ * all callsites. This is a minimum of 1 when a function is
+ * pointwise. If the footprint cannot be calculated, returns
+ * UNDEFINED_FOOTPRINT_SIZE. */
+unsigned calculate_footprint_size(Function f, CallGraph &cg) {
+    vector<Interval> bounds = get_function_bounds(f, cg);
+    if (bounds.size() == 0) {
+        // Unable to calculate function bounds.
+        return UNDEFINED_FOOTPRINT_SIZE;
+    }
+    Expr footprint = 1;
+    for (Interval i : bounds) {
+        internal_assert(i.min.defined() && i.max.defined());
+        Expr diff = i.max - i.min + 1;
+        footprint *= diff;
+    }
+    const int *result = as_const_int(simplify(footprint));
+    if (result) {
+        return (unsigned)*result;
+    } else {
+        return UNDEFINED_FOOTPRINT_SIZE;
+    }
+}
+
+} // end anonymous namespace
+
+void ComputeRootAllStencils::apply(Func root) {
+    // Construct a callgraph for the pipeline.
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    for (Function f : all_functions) {
+        unsigned footprint = calculate_footprint_size(f, cg);
+        if (footprint > 1) {
+            Func wrapper(f);
+            wrapper.store_root().compute_root();
+        }
+    }
+}
+
+void ParallelizeOuter::apply(Func root) {
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    all_functions.push_back(root.function());
+    for (Function f : all_functions) {
+        if (!f.schedule().compute_level().is_inline()) {
+            Func wrapper(f);
+            Dim outer = f.schedule().dims()[f.schedule().dims().size() - 1];
+            Var v(outer.var);
+            wrapper.parallel(v);
+        }
+    }
+}
+
+void VectorizeInner::apply(Func root) {
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    all_functions.push_back(root.function());
+    for (Function f : all_functions) {
+        if (!f.schedule().compute_level().is_inline()) {
+            Func wrapper(f);
+            Dim inner = f.schedule().dims()[0];
+            Var v(inner.var);
+            unsigned factor = 128 / f.output_types()[0].bits;
```

Comment:
This number should probably be 256 so that it hits avx. On targets with 128-bit vectors the backend should slice things up cleanly.